### ISSUE TITLE
Incorporate SecurityManager into Spacelift

### ISSUE THREAD

**Issue by smiklosovic** [arquillian/arquillian-spacelift#28]
It could be done simply just on CommandTool or on Spacelift object.

Examples:

```
Spacelift.task(CommandTool.class)
    .programName("echo")
    .paramaters("hello world")
    .execute()
    .withPermission(Permisson p) // by default used java.security.AllPermission
    .withSecurityManager(SecurityManager sm) // used default when not set
    .await()
```

on Spacelift object

```
Spacelift.withPermission(Permission p).withSecurityManager(SecurityManager sm)
    .task(CommandTool.class)
    .programName("echo")
    .paramaters("hello world")
    .execute()
    .await()
```

```
Spacelift.security(Permission p)
Spacelift.security(Permission p, SecurityManager sm)
```

I would go for adding it to Task class instead of Spacelift because you could specify security managers and permissions on every (possibly chained) task and not globally on a Spacelift object.

**Comment by kpiwko** [comment]
That would work only if we make sure that a task does not execute other task internally. Any ideas how to ensure that?

**Comment by smiklosovic** [comment]
heh, I didnt think about that. We could kind of hack it, something like counting the number of process() method invocations before eventual then(). process() invocation would increase counter by one, then() would decrease it. So if you ever reach then() with non zero counter, it means you had to invoke another process() while you are inside a process().

However I do not know what about the case when you are chaining it inside a process() method.

**Comment by kpiwko** [comment]
Counter would not work reliably. Think about parallel executions and such. I would say that we can make security manager flat - that is keep it only for `Spacelift` object and inherit it all subsequent calls.

Do you have a real example what your trying to solve? It might help to define API better.

**Comment by smiklosovic** [comment]
I agree that it is clunky solution so doing it on Spacelift level would be better.

I do not have any particular problem to solve with this. It is a good question if we really want to implement it. However when you look at it plainly practically, I do not see the reason why not.

For example, your Task could read some properties and you do not want to allow them being read. Or some networking stuff, writing to files and so on ...
